### ISSUE TITLE
How do I mix continuous and categorical inputs?

### ISSUE THREAD

**Issue by cinjon** [rusty1s/pytorch_geometric#2044]
## ❓ Questions & Help

I have a graph where each node has some subset of common features. These features could be continuous or categorical. If they are continuous, I would like to use the (normalized) value as the input. If they are categorical, then I'd like for the model to learn an embedding representing that input. 

I am having trouble understanding how exactly to deliver the input to this library. Assuming I'm using GCNConv, should the num_node_features input size be the total number of possible input variables (sum of the cardinality of the continuous and categorical sets)? Should it instead be the cardinality of the continuous set plus embedding_dim * cardinality of the categorical set? The latter would then require first getting the embeddings of the input (using nn.Embedding).

Currently, I am building these attributes, putting them into a geometric.Data object, and then applying Batch.from_data_list(<list of geometric.Data objects>) to get the batch. Do I need to fill in _every_ categorical slot to this in order for batching to work correctly or can I give the Data objects a keyword corresponding to each category?

**Comment by rusty1s** [comment]
It's the user's choice how the encode features into a feature vector. Then, `GCNConv` expects the dimensionality of this vector as the `in_channels` attribute.

If you have both continuous and categorical features, a good choice is the following:
1. As you said, for continuous ones, you can use them as they are present in the data (it's always good to normalize them nonetheless)
2. For categorical ones, I suggest them to encode them into a one-hot vector (which is equivalent to encode them as trainable embeddings, but moves optimization outside of the data).

As an example:
```python
x1 = (x1 - x1.mean(dim=0)) / x1.std(dim=0)  # Continues features of shape [num_nodes, d1]
x2 = ... # Categorical features of shape [num_nodes, d2]
xs = []
for i in range(d2):
    xs += [torch.nn.functional.one_hot(x2[:, i], num_classes=...).to(torch.float)]
x = torch.cat([x1] + x2, dim=-1)

conv = GCNConv(x.size(-1), 256)
```

**Comment by cinjon** [comment]
Thanks for the prompt reply! Three ensuing questions:

1. In your example, you build xs but then don't use it. Did you mean to do ```x = torch.cat([x1] + xs, ...)```? I'm assuming so because otherwise the GCN is operating on the continuous values and the categorical indices. 
2. If the GCN is operating on the continuous values and the categorical one-hots, then is there some place here where an embedding dictionary comes in that I'm not seeing? I would expect that GCNConv would be operating on the embeddings rather than the one-hots.
3. As I understand, the input size to GCNConv can't be variable. However, the nodes all have some subset of the attributes and never actually the full breadth of them. Is the idea here that we expand each node's attribute list to give them all each of the attributes, copiously use nullifying values (like 0 for continuous and a "Nothing" slot for categorical)?

**Comment by rusty1s** [comment]
1. Fixed in the version above.
2. The embedding automatically comes into place in the first `Linear` layer that takes in those features. This weight matrix will represent your embedding for one-hot encoded values.
3. Yes, you need to fill those out with zeros (doable via `num_classes=...` in `one_hot`).

**Comment by cinjon** [comment]
Amazing, thanks! Much appreciated :).
